### ISSUE TITLE
CPU conv2d backward produces incorrect gradients for non-contiguous kernels

### ISSUE THREAD

**Issue by EthanAlmloff** [huggingface/candle#3735]
## Summary

On CPU, `conv2d` backward can produce incorrect gradients when the convolution kernel is non-contiguous.

A transposed kernel and its contiguous copy represent the same logical values, so they should produce identical forward results and gradients. Instead, the gradients differ.

## Reproduction

```rs
use candle_core::Var;
use candle_core::{test_utils, Device, Var};

let device = &Device::Cpu;
let input_data: Vec<f32> = (1..=18).map(|value| value as f32).collect();
let kernel = Var::from_slice(
    &(1..=16).map(|value| value as f32).collect::<Vec<_>>(),
    (2, 2, 2, 2),
    device,
)?;
let non_contiguous_kernel = kernel.transpose(0, 1)?;

let input = Var::from_slice(&input_data, (1, 2, 3, 3), device)?;
let loss = input
    .conv2d(&non_contiguous_kernel, 0, 1, 1, 1)?
    .sqr()?
    .sum_all()?;
let grads = loss.backward()?;
let grad_input = grads.get(&input).unwrap();
let grad_kernel = grads.get(&kernel).unwrap();

let expected_input = Var::from_slice(&input_data, (1, 2, 3, 3), device)?;
let expected_kernel = Var::from_slice(
    &(1..=16).map(|value| value as f32).collect::<Vec<_>>(),
    (2, 2, 2, 2),
    device,
)?;
let expected_loss = expected_input
    .conv2d(&expected_kernel.transpose(0, 1)?.contiguous()?, 0, 1, 1, 1)?
    .sqr()?
    .sum_all()?;
let expected_grads = expected_loss.backward()?;
let expected_grad_input = expected_grads.get(&expected_input).unwrap();
let expected_grad_kernel = expected_grads.get(&expected_kernel).unwrap();

assert_eq!(
    test_utils::to_vec1_round(&grad_input.flatten_all()?, 4)?,
    test_utils::to_vec1_round(&expected_grad_input.flatten_all()?, 4)?,
);
assert_eq!(
    test_utils::to_vec1_round(&grad_kernel.flatten_all()?, 4)?,
    test_utils::to_vec1_round(&expected_grad_kernel.flatten_all()?, 4)?,
);
Ok(())
```

## Expected behavior

The gradients for input and kernel should match the gradients from the equivalent contiguous-kernel computation.

## Actual behavior

The CPU gradients differ when the kernel is a non-contiguous transpose view.

**Comment by EthanAlmloff** [comment]
This was already fixed with newer versions of Candle than I was using.
